mirascope/base/extractors.py

```diff
@@ -121,32 +123,57 @@ def _extract(
             AttributeError: if there is no tool in the call creation.
             ValidationError: if the schema cannot be instantiated from the completion.
         """
-        kwargs, return_tool = self._setup(tool_type, kwargs)
 
-        temp_call = self._generate_temp_call(call_type)
+        def _extract_attempt(
+            call_type: Type[BaseCallT],
+            tool_type: Type[BaseToolT],
+            error_messages: dict[str, Any],
+            **kwargs: Any,
+        ) -> ExtractedTypeT:
+            kwargs, return_tool = self._setup(tool_type, kwargs)
+
+            temp_call = self._generate_temp_call(call_type, error_messages)
+            obj = temp_call(
+                **self.model_dump(exclude={"extract_schema"}),
+            )
+            print(obj.messages())
+            response = obj.call(**kwargs)
+            try:
+                extracted_schema = self._extract_schema(
+                    response.tool, self.extract_schema, return_tool, response=response
+                )
+                if extracted_schema is None:
+                    raise AttributeError("No tool found in the completion.")
+                return extracted_schema
+            except (AttributeError, ValueError, ValidationError):
+                raise
 
-        response = temp_call(**self.model_dump(exclude={"extract_schema"})).call(
-            **kwargs
-        )
+        if isinstance(retries, int):
+            retries = Retrying(stop=stop_after_attempt(retries))
         try:
-            extracted_schema = self._extract_schema(
-                response.tool, self.extract_schema, return_tool, response=response
-            )
-            if extracted_schema is None:
-                raise AttributeError("No tool found in the completion.")
-            return extracted_schema
-        except (AttributeError, ValueError, ValidationError) as e:
-            if retries > 0:
-                logging.info(f"Retrying due to exception: {e}")
-                # TODO: include failure in retry prompt.
-                return self._extract(call_type, tool_type, retries - 1, **kwargs)
-            raise  # re-raise if we have no retries left
+            error_messages: dict[str, Any] = {}
+            for attempt in retries:
+                with attempt:
+                    try:
+                        extraction = _extract_attempt(
+                            call_type, tool_type, error_messages, **kwargs
+                        )
+                    except (AttributeError, ValueError, ValidationError) as e:
+                        error_messages[str(e)] = None
+                        if (
+                            self.call_params.logfire or self.call_params.logfire_async
+                        ):  # pragma: no cover
```

Comment:
does this cover everything inside the conditional?

Reply:
yes